h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1131,13 +1186,29 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
 
           // Gradient wrt x_i is matrix product \grad L_{i,j}(x_i * Y_j, A_{i,j}) * Y_j'
           double[] weight = _lossFunc[j].mlgrad(xy, (int) a[j]);
-          double[][] ysub = _yt.getCatBlock(j);
+ //         double[][] ysub = _yt.getCatBlock(j);
           for (int k = 0; k < _ncolX; k++) {
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+            for (int c = 0; c < weight.length; c++) {
+ //             grad[k] += cweight * weight[c] * ysub[k][c];
+              int cidx = _yt.getCatCidx(j, c);
+              double archtypevalues = 0;
+              if (_yt._transposed)
+                archtypevalues = _yt._archetypes[cidx][k];
```

Comment:
Tomas:

We need to calculate the Gram matrix which is Y_transpose * Y or the other way round.  So, sometimes, we will have our archetypes in normal form and sometimes in the transposed form.  I do not fully understand this thingy yet and is hesitate to make the big change.  

However, I can make small tweaks without breaking the algorithm.  Once I understand fully what is going on, I will make better changes that you would prefer.  Please stay tune.

I would like to hear your comments even if it is not needed for this update.


Thanks,
Wendy